In [1]:
import pandas as pd

In [2]:
#read in our data
df_full = pd.read_csv('data/2018_missoula_marathon_full.csv')
df_half = pd.read_csv('data/2018_missoula_marathon_half.csv')

In [3]:
df_full.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1034 entries, 0 to 1033
Data columns (total 12 columns):
Place        1034 non-null int64
Name         1034 non-null object
City         1012 non-null object
Bib_No       1034 non-null int64
Age          1034 non-null int64
Gender       1034 non-null object
Age_Group    1034 non-null object
Chip_Time    1034 non-null object
Gun_Time     1034 non-null object
Chip_Diff    1028 non-null object
Pace         1034 non-null object
Race         1034 non-null object
dtypes: int64(3), object(9)
memory usage: 97.0+ KB


In [4]:
df_half.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2659 entries, 0 to 2658
Data columns (total 12 columns):
Place        2659 non-null int64
Name         2659 non-null object
City         2645 non-null object
Bib_No       2659 non-null int64
Age          2658 non-null float64
Gender       2659 non-null object
Age_Group    2659 non-null object
Chip_Time    2659 non-null object
Gun_Time     2659 non-null object
Chip_Diff    2647 non-null object
Pace         2659 non-null object
Race         2659 non-null object
dtypes: float64(1), int64(2), object(9)
memory usage: 249.4+ KB


Looking at the info, it appears that we have issues with missing 'City' fields and missingt 'Chip_Diff' fields.

All missing 'City' entries are for race pacers. Let's make an assumption that they are all from Missoula. The missing 'Chip_Diff' entries occur when the chip time and the gun time are the same, so let's replace those with zeros.

There's one more issue to deal with. We're missing an age for one male half marathon runner (bib number 9384). That runner is listed as 'Unknown Participant' and is missing other details as well. Let's drop that record.

In [5]:
#first let's drop our rogue runner
df_half = df_half[df_half.Bib_No != 9384]
df_half.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2658 entries, 0 to 2658
Data columns (total 12 columns):
Place        2658 non-null int64
Name         2658 non-null object
City         2645 non-null object
Bib_No       2658 non-null int64
Age          2658 non-null float64
Gender       2658 non-null object
Age_Group    2658 non-null object
Chip_Time    2658 non-null object
Gun_Time     2658 non-null object
Chip_Diff    2646 non-null object
Pace         2658 non-null object
Race         2658 non-null object
dtypes: float64(1), int64(2), object(9)
memory usage: 270.0+ KB


In [6]:
df_half[df_half.City.isna()]

Place                        Name City  Bib_No   Age Gender  \
87       88        Kate Johnson - Pacer  NaN    3400  26.0      F   
519     520     Ilana Abrahamson -Pacer  NaN    2107  42.0      F   
520     521     Sara McAllister - Pacer  NaN    3819  35.0      F   
630     631         Rachel Toor - Pacer  NaN    7029  56.0      F   
943     944     Theneshia Glaze - Pacer  NaN    3011  41.0      F   
978     979       Rebecca Garner -Pacer  NaN    2971  52.0      F   
1903     51          Dillon May - Pacer  NaN    3809  21.0      M   
1904     52     Nicolas Composto -Pacer  NaN    2578  25.0      M   
1997    145         Scott Davis - Pacer  NaN    2689  26.0      M   
2184    332     Danny Tenenbaum - Pacer  NaN    4799  32.0      M   
2185    333  David Schmetterling -Pacer  NaN    4529  46.0      M   
2319    467       Vo Von Sehlen - Pacer  NaN    7031  53.0      M   
2341    489      Keith Fichtner - Pacer  NaN    2886  42.0      M   

          Age_Group    Chip_Time    Gun_Time Chip_Diff     Pace  Race  
87     14/201:25-29   1:45:16.99  1:45:26.48   0:09.49   8:02/M  half  
519    68/211:40-44   2:09:46.25  2:10:25.72   0:39.47   9:54/M  half  
520   103/293:35-39   2:09:46.70  2:10:26.13   0:39.43   9:54/M  half  
630    24/152:55-59   2:14:36.70  2:15:48.73   1:12.03  10:16/M  half  
943   117/211:40-44   2:29:50.89  2:31:52.41   2:01.52  11:26/M  half  
978    68/173:50-54   2:31:05.50  2:33:06.82   2:01.32  11:32/M  half  
1903    12/32:20-24   1:29:38.56  1:29:42.16   0:03.60   6:51/M  half  
1904     5/68:25-29   1:29:38.79  1:29:42.21   0:03.42   6:51/M  half  
1997    15/68:25-29   1:45:17.19  1:45:26.60   0:09.41   8:02/M  half  
2184    56/97:30-34   2:00:16.96  2:00:39.64   0:22.68   9:11/M  half  
2185    34/91:45-49   2:00:17.55  2:00:40.14   0:22.59   9:11/M  half  
2319    39/69:50-54   2:14:34.52  2:15:49.42   1:14.90  10:16/M  half  
2341    53/84:40-44   2:17:38.43  2:19:03.49   1:25.06  10:30/M  half

In [7]:
df_half.City = df_half.City.fillna('Missoula MT US')

In [8]:
df_half[df_half.City.isna()]

Empty DataFrame
Columns: [Place, Name, City, Bib_No, Age, Gender, Age_Group, Chip_Time, Gun_Time, Chip_Diff, Pace, Race]
Index: []

In [9]:
df_half[df_half.Chip_Diff.isna()]

Place               Name                   City  Bib_No   Age Gender  \
707     708     Emily Cheroske        Missoula  MT US    5182  22.0      F   
1007   1008     Claire Rappold        Missoula  MT US    4333  25.0      F   
1012   1013      Janice Loudon   Overland Park  KS US    3703  58.0      F   
1659   1660    Marybeth Healey   Coeur D Alene  ID US    3160  57.0      F   
1788   1789         Dawn Clark           Butte  MT US    2546  59.0      F   
1808   1809      Amanda Ledoux  Spokane Valley  WA US    3644  44.0      F   
1809   1810      Celina Taylor         Spokane  WA US    4788  42.0      F   
1844   1845        Alina McCue        Missoula  MT US    5185  39.0      F   
1889     37      Connor Morris    Port Orchard  WA US    5298  20.0      M   
1923     71  Lucca Wellenstein        Missoula  MT US    4993  16.0      M   
1974    122         Ian Curtis        Missoula  MT US    2649  16.0      M   
2110    258       Tom Frandsen            Lolo  MT US    2930  34.0      M   

          Age_Group    Chip_Time    Gun_Time Chip_Diff     Pace  Race  
707    66/121:20-24   2:18:57.49  2:18:57.49       NaN  10:36/M  half  
1007  121/201:25-29   2:33:05.99  2:33:05.99       NaN  11:41/M  half  
1012   62/152:55-59   2:33:41.21  2:33:41.21       NaN  11:44/M  half  
1659  128/152:55-59   3:32:18.86  3:32:18.86       NaN  16:12/M  half  
1788  146/152:55-59   4:00:36.85  4:00:36.85       NaN  18:22/M  half  
1808  202/211:40-44   4:09:38.16  4:09:38.16       NaN  19:03/M  half  
1809  203/211:40-44   4:09:38.62  4:09:38.62       NaN  19:03/M  half  
1844  293/293:35-39   4:31:55.15  4:31:55.15       NaN  20:45/M  half  
1889    11/32:20-24   1:26:38.11  1:26:38.11       NaN   6:37/M  half  
1923    16/45:15-19   1:33:38.00  1:33:38.00       NaN   7:09/M  half  
1974    25/45:15-19   1:41:59.79  1:41:59.79       NaN   7:47/M  half  
2110    40/97:30-34   1:55:03.88  1:55:03.88       NaN   8:47/M  half

In [10]:
df_half.Chip_Diff = df_half.Chip_Diff.fillna(0)
df_half[df_half.Chip_Diff.isna()]

Empty DataFrame
Columns: [Place, Name, City, Bib_No, Age, Gender, Age_Group, Chip_Time, Gun_Time, Chip_Diff, Pace, Race]
Index: []

In [11]:
#let's fix those same issues with the full results.
df_full.City = df_full.City.fillna('Missoula MT US')
df_full.Chip_Diff = df_full.Chip_Diff.fillna(0)

In [12]:
df_full.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1034 entries, 0 to 1033
Data columns (total 12 columns):
Place        1034 non-null int64
Name         1034 non-null object
City         1034 non-null object
Bib_No       1034 non-null int64
Age          1034 non-null int64
Gender       1034 non-null object
Age_Group    1034 non-null object
Chip_Time    1034 non-null object
Gun_Time     1034 non-null object
Chip_Diff    1034 non-null object
Pace         1034 non-null object
Race         1034 non-null object
dtypes: int64(3), object(9)
memory usage: 97.0+ KB


In [13]:
df_half.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2658 entries, 0 to 2658
Data columns (total 12 columns):
Place        2658 non-null int64
Name         2658 non-null object
City         2658 non-null object
Bib_No       2658 non-null int64
Age          2658 non-null float64
Gender       2658 non-null object
Age_Group    2658 non-null object
Chip_Time    2658 non-null object
Gun_Time     2658 non-null object
Chip_Diff    2658 non-null object
Pace         2658 non-null object
Race         2658 non-null object
dtypes: float64(1), int64(2), object(9)
memory usage: 270.0+ KB


Ok. We have our missing/messy values worked out. Next, work our our columns.

The 'Age_Group' field is holding both the runners place out of the total for that age group and the label. For example:

66/121:20-24 - The runner was number 66 out of 121 runners in the 20-24 age group.

Let's drop the total number in the group. We can always calculate that if we need it. And let's split the data into two colums one for place and one for group.

In [14]:
df_half['Age_Group_Place'], df_half['Age_Group'] = df_half.Age_Group.str.split(':', 1).str
df_half['Age_Group_Place'], _ = df_half.Age_Group_Place.str.split('/', 1).str
df_half.head()

Place                 Name                     City  Bib_No   Age Gender  \
0      1        Makena Morley           Bigfork  MT US    2051  21.0      F   
1      2     Misiker Demessie  Colorado Springs  CO US    2060  31.0      F   
2      3         Kelsi Lasota        Estes Park  CO US    2052  22.0      F   
3      4  Elizabeth Wasserman            Goshen  NY US    2056  23.0      F   
4      5     Andrea Masterson           Seattle  WA US    2053  21.0      F   

  Age_Group    Chip_Time    Gun_Time Chip_Diff    Pace  Race Age_Group_Place  
0     20-24   1:15:50.45  1:15:51.79   0:01.34  5:47/M  half               1  
1     30-34   1:18:03.69  1:18:05.58   0:01.89  5:57/M  half               1  
2     20-24   1:18:41.68  1:18:43.68   0:02.00  6:00/M  half               2  
3     20-24   1:19:18.71  1:19:20.92   0:02.21  6:03/M  half               3  
4     20-24   1:21:38.93  1:21:40.32   0:01.39  6:14/M  half               4

In [15]:
df_full['Age_Group_Place'], df_full['Age_Group'] = df_full.Age_Group.str.split(':', 1).str
df_full['Age_Group_Place'], _ = df_full.Age_Group_Place.str.split('/', 1).str
df_full.head()

Place            Name                City  Bib_No  Age Gender Age_Group  \
0      1     Keely Baker  Great Falls  MT US      52   26      F     25-29   
1      2  Trisha Drobeck     Missoula  MT US      51   38      F     35-39   
2      3    Chelsea Dana     Billings  MT US     355   35      F     35-39   
3      4       Evie Tate     Missoula  MT US    1211   23      F     20-24   
4      5      Erin Forde     Missoula  MT US      56   25      F     25-29   

     Chip_Time    Gun_Time Chip_Diff    Pace  Race Age_Group_Place  
0   2:51:46.12  2:51:48.01   0:01.89  6:33/M  full               1  
1   2:53:28.28  2:53:30.08   0:01.80  6:37/M  full               1  
2   3:01:37.43  3:01:43.04   0:05.61  6:56/M  full               2  
3   3:09:32.24  3:09:45.38   0:13.14  7:14/M  full               1  
4   3:09:37.84  3:09:41.66   0:03.82  7:14/M  full               2